Add more metadata into output image

mfojtik opened the pull request:
This PR will make s2i store more metadata about the source code the image was built from.  This will help debugging problems as you can later, in runtime identify what revision of the app is the container running.

mfojtik commented:
@bparees @soltysh PTAL (this is not urgent, just good to have for investigating problems)

bparees commented:
@mfojtik https://trello.com/c/cd1S05XE/596-add-commit-author-info-labels-to-built-images

please make sure whatever you're doing here aligns with the needs of that card.

mfojtik commented:
yes i can add all of them
On Jun 19, 2015 8:24 PM, "Ben Parees" <notifications@github.com> wrote:

> @mfojtik <https://github.com/mfojtik>
> https://trello.com/c/cd1S05XE/596-add-commit-author-info-labels-to-built-images
>
> please make sure whatever you're doing here aligns with the needs of that
> card.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/source-to-image/pull/215#issuecomment-113597655>
> .
>

soltysh commented on pkg/api/types.go:
```diff
@@ -121,3 +121,26 @@ type InstallResult struct {
+
+	// The GIT commit id
+	// This information is available in the image as 'io.openshift.build.commit.id' LABEL
+	ID string
```
Pls rename this to to `CommitID`, since `ID` is too vague IMHO. I assume the actual output data will match the name of this field, right?

bparees replied:
+1 commitid.

bparees replied:
behavior for these fieids:

ref is a commit.  commitid will match ref
ref is a branch.  commitid will be a commit in that branch
ref is a tag.  commitid will be the commit pointed to by the tag.

soltysh commented on pkg/build/strategies/sti/sti.go:
```diff
@@ -218,6 +219,17 @@ func (b *STI) PostExecute(containerID, location string) error {
+	labels := map[string]string{}
+	if b.sourceInfo != nil {
+		labels = map[string]string{
+			"io.openshift.build.commit.author":     b.sourceInfo.Author,
```
I've talked with @bparees about that, we've agreed to have the prefix configurable. By default it'll be `io.s2i...` and in OpenShift it'll be `io.openshift...`

mfojtik replied:
i don't want to have this configurable because it makes things much harder for no good reasons.

soltysh replied:
k

soltysh commented on pkg/build/strategies/sti/sti.go:
```diff
@@ -218,6 +219,17 @@ func (b *STI) PostExecute(containerID, location string) error {
+			"io.openshift.build.commit.date":       b.sourceInfo.Date,
+			"io.openshift.build.commit.id":         b.sourceInfo.ID,
+			"io.openshift.build.commit.ref":        b.sourceInfo.Ref,
+			"io.openshift.build.source-repository": b.sourceInfo.Location,
```
I'm a little bit concerned that source-repository here isn't the right naming, since you can build from local directory, which is not a source-repository (btw. how should we handle relative vs absolute paths then? always expand to absolute path?) One other issue is context directory, that is missing in here and it **has** to be here imho.

mfojtik replied:
@soltysh local directory **must be** source repository

bparees replied:
still, calling it source-path or source-location would future proof us.

soltysh replied:
Ok, still context directory is something we need here.

bparees replied:
true.

mfojtik commented:
Updated this PR to include the LABELs:

```console
I0622 16:00:28.329682 20941 docker.go:471] Committing container with config: {Hostname: Domainname: User: Memory:0 MemorySwap:0 CPUShares:0 CPUSet: AttachStdin:false AttachStdout:false AttachStderr:false PortSpecs:[] ExposedPorts:map[] Tty:false OpenStdin:false StdinOnce:false Env:[] Cmd:[/usr/local/sti/run] DNS:[] Image: Volumes:map[] VolumesFrom: WorkingDir: Entrypoint:[] NetworkDisabled:false SecurityOpts:[] OnBuild:[] Labels:map[io.openshift.build.commit.author:Michal Fojtik <mfojtik@redhat.com> io.openshift.build.commit.date:Thu May 7 17:14:28 2015 +0200 io.openshift.build.commit.id:master io.openshift.build.commit.ref:7ee9858 io.openshift.build.source-repository:file:///home/mfojtik/sandbox/ruby]}
```

mfojtik commented:
So the current set of labels is this:
```json
        "Labels": {
            "io.openshift.build.commit.author": "Michal Fojtik \u003cmfojtik@redhat.com\u003e",
            "io.openshift.build.commit.date": "Thu May 7 17:14:28 2015 +0200",
            "io.openshift.build.commit.id": "master",
            "io.openshift.build.commit.ref": "7ee9858",
            "io.openshift.build.source-repository": "file:///home/mfojtik/sandbox/ruby",
            "io.s2i.scripts-url": "image:///usr/local/sti",
            "k8s.io/description": "Platform for building and running Ruby 2.0 applications",
            "k8s.io/display-name": "Ruby 2.0",
            "openshift.io/expose-services": "8080:http",
            "openshift.io/tags": "builder,ruby,ruby20"
        },

```

mfojtik commented:
[test]

No idea why the Travis failed, I can build this just fine locally.

bparees commented on pkg/api/types.go:
```diff
@@ -121,3 +121,26 @@ type InstallResult struct {
+
+	// Author of the commit
+	// This information is available in the image as 'io.openshift.build.commit.author' LABEL
+	Author string
```
add the commit comment as a field too.

mfojtik replied:
@bparees the commit message can be pretty long... are you sure you want to have it as label? I mean if you have source location and GIT ref, then you can just fetch the commit and get the message... Alternatively we can truncate the message to some reasonable length, or just take the first line.

mfojtik commented:
@bparees @soltysh all comments addressed

mfojtik commented:
@bparees @soltysh the only thing that bothers me here is the label format inconsistency... The final image labels looks like this now:

```json
            "io.openshift.build.commit.author": "Michal Fojtik \u003cmfojtik@redhat.com\u003e",
            "io.openshift.build.commit.date": "Thu May 7 17:14:28 2015 +0200",
            "io.openshift.build.commit.id": "master",
            "io.openshift.build.commit.ref": "7ee9858",
            "io.openshift.build.source-location": "file:///home/mfojtik/sandbox/ruby",
            "io.s2i.scripts-url": "image:///usr/local/sti",
            "k8s.io/description": "Platform for building and running Ruby 2.0 applications",
            "k8s.io/display-name": "Ruby 2.0",
            "openshift.io/expose-services": "8080:http",
            "openshift.io/tags": "builder,ruby,ruby20"
```

I think we have to make this consistent, either in OpenShift STI images (if it is not too late for that) or in STI.

mfojtik commented:
@bparees @soltysh also note that the 'display-name' is not quite correct here... the displayName should be the application name and not the builder image name.

soltysh commented:
Isn't display-name inherited from the builder itself?

soltysh commented on pkg/api/types.go:
```diff
@@ -121,3 +121,32 @@ type InstallResult struct {
+
+	// SourceLocation contains a valid URL to the original repository.
+	// The output image will contain this information as 'io.openshift.build.source-location' label.
+	SourceLocation string
```
If the struct is called `SourceInfo`, having this field named just `Location` would be enough, the same applies to `SourceContextDir` -> `ContextDir`.

mfojtik replied:
@soltysh my intent was to match them with the label names

mfojtik replied:
fixed.

soltysh commented:
A few problems:
1) I did:
```
sti build https://github.com/openshift/sti-python.git --context-dir=3.3/test/setup-test-app/ openshift/python-33-centos7 python-sample-app
```
and got in result:
```
            "io.openshift.build.commit.author": "",
            "io.openshift.build.commit.date": "",
            "io.openshift.build.commit.id": "",
            "io.openshift.build.commit.ref": "",
            "io.openshift.build.source-context-dir": "3.3/test/setup-test-app/",
            "io.openshift.build.source-location": "",
            "io.s2i.scripts-url": "image:///usr/local/sti",
            "k8s.io/description": "Platform for building and running Python 3.3 applications",
            "k8s.io/display-name": "Python 3.3",
            "openshift.io/expose-services": "8080:http",
            "openshift.io/tags": "builder,python,python33"
```

I didn't specify any ref, nor commit id, still I'd expect both to contain master and latest commit id respectively, along with author and date. Why the source-location is also empty?


2) I also did this:
```
sti build /tmp/xxx openshift/python-33-centos7 python-sample-app
```
where `/tmp/xxx` contains the files from above context dir and got no info about build. Which is OK, but also means we don't require source to be in any repository, which you told me we do. See [this line](https://github.com/openshift/source-to-image/blob/master/pkg/git/clone.go#L59) which handles copying sources if no valid git clone is specified [this check](https://github.com/openshift/source-to-image/blob/master/pkg/git/clone.go#L22).


3) Finally I did this:
```
sti build 3.3/test/setup-test-app/ openshift/python-33-centos7 python-sample-app
```
from directory where [sti-python](https://github.com/openshift/sti-python) is cloned, and again got no info about build, but I'd expect it should be smart enough to get them or at least try to get them, IMHO.


Points 1 and 3 are the ones that worries me.

mfojtik commented:
@soltysh good catch... the 1) and 3) cases we remove the source directory so we can't execute git commands there... I fixed this to gather the info before we remove stuff.

bparees commented:
@mfojtik @soltysh 
yeah short of introducing a new "--name" flag to S2I, the display-name is going to be inherited...  maybe it's worth tagging "app" onto the end of the existing display name, to at least identify it a little, pending the decision to introduce such a flag?  (and to have a default behavior if the user doesn't specify the flag?)

mfojtik commented:
@bparees @soltysh I will do what you are proposing but also would be nice to offer '--name' and '--description'. If they are set then they override the builder image tags.

soltysh commented:
How about taking for the app name the image name (the last part after / sign, without the tag part of course) if none is specified?

mfojtik commented:
@soltysh we are talking about the 'display name' here. Your proposal is more a 'name' of the build config :-)

bparees commented:
@mfojtik no i think @soltysh is saying:

    $ sti build https://myrepo.github.com mybuilderimage myorg/myappimage

should result in the display name being "myappimage"

but i have to say that seems a little pointless... you already know the image name so if you're going to make the display name be the image name, you don't need the display name field to be set at all.

soltysh commented:
@bparees +1, was about to write that down :wink:

soltysh commented:
@bparees well, either we'll clear the display-name when nothing is passed, which is necessary because we don't want to have Ruby 2.0 there, or give there some meaningful name, in which case falling back to output image was quite natural imho. Besides the image name is one thing and diplay-name something different, unless the logic will take the image name if no display-name is provided in that case, that makes sense.

bparees commented:
@mfojtik I can't find where my comment went but for commit message first sentence is fine with me.

mfojtik commented:
@bparees i'm saying that 'myappimage' is not a good 'display name' but a good 'name' (eg. for buildConfig)

bparees commented:
@mfojtik for the commit message, search for \n or 80 chars, whichever comes first.
if you hit 80 chars, take the first 77 and append "..."

bparees commented:
i'm fine with either defaulting to the imagename for display-name, or taking the existing display-name and appending "app"  (keep in mind not all builder images will have a display-name label either)

mfojtik commented:
(the label discussion is moved here: https://github.com/openshift/origin/issues/3414)

mfojtik commented:
@bparees @soltysh ready for next round of reviews.... summary of changes:

* Added `--application-name` and `--description` options to `sti build`. They are both optional and the `--application-name` defaults to output image name
* Added `io.openshift.s2i.build.commit.message` (truncated to 80 chars)
* Added ` source-context-dir` label (when the contextDir is set)
* When building from random folder, we attempt to read the metadata anyway (what if the folder is GIT repo after all? ;-)... when we fail, we tell it
* More plumbing around namespaces

rhcarvalho commented on pkg/test/git.go:
```diff
@@ -33,3 +35,11 @@ func (f *FakeGit) Checkout(repo, ref string) error {
+	return &api.SourceInfo{
+		Ref:      "master",
+		CommitID: "1bf4f04",
+		Location: "file;//foo",
```
`file:///foo`?

soltysh commented on pkg/git/git.go:
```diff
@@ -90,6 +93,28 @@ func (h *stiGit) Checkout(repo, ref string) error {
+	}
+	return &api.SourceInfo{
+		Location: git("config", "--get", "remote.origin.url"),
+		CommitID: git("rev-parse", "--abbrev-ref", "HEAD"),
```
This returns the ref actually, and the command in ref returns commitID, you need to switch them.

soltysh commented on pkg/api/types.go:
```diff
@@ -2,8 +2,20 @@ package api
 // Config contains essential fields for performing build.
 type Config struct {
+	// DisplayName is a result image display-name label. This defaults to the
+	// result image name.
```
s/result/output

soltysh commented on pkg/api/types.go:
```diff
@@ -121,3 +133,36 @@ type InstallResult struct {
+// SourceInfo stores information about the source code
+type SourceInfo struct {
+	// Ref represents a commit SHA-1, valid GIT branch name or a GIT tag
+	// The output image will contain this information as 'io.openshift.build.commit.ref' label.
```
Actually now it's `io.openshift.s2i.build.commit.ref`, rewrite it to say `DefaultNamespace + 'commit.ref'`.
This also applies to all other `SourceInfo` fields.

mfojtik replied:
@soltysh I hate referring to const in comments ;-)

soltysh replied:
Your choice :stuck_out_tongue:

soltysh commented on pkg/git/clone.go:
```diff
@@ -44,17 +44,30 @@ func (c *Clone) Download(config *api.Config) error {
-		return errors.NewSourcePathError(path)
+		return nil, errors.NewSourcePathError(path)
+	}
+	if err := c.Copy(path, targetSourceDir); err != nil {
```
I'd try getting info before doing copy, because if I do (my 3rd test case):
```
sti build 3.3/test/setup-test-app/ openshift/python-33-centos7 python-sample-app
```
From inside sti-python this could return valid info, assuming you do that in the source (represented by `path` here), but after copying all the git info is lost.

soltysh replied:
Otherwise, you will never be able to get decent info about that.

mfojtik replied:
@soltysh when using non-git repo or just a subfolder without contextDir then we should not gather **any** info at all and don't set any labels.

soltysh replied:
Spoke on IRC, I'm ok with current solution.

soltysh commented:
@mfojtik a couple of small nits :wink: and you're almost done.

mfojtik commented:
@soltysh thanks for reviewing this, I fixed all comments (except the two). I will fix the label comments as follow up when we decide what labels format we are going to use.

soltysh commented:
LGTM, will wait with merge for @bparees approval

bparees commented on pkg/api/types.go:
```diff
@@ -2,8 +2,20 @@ package api
 
+const (
+	DefaultNamespace    = "io.openshift.s2i."
+	KubernetesNamespace = "k8s.io/"
```
thought @smarterclayton said k8s was using docker style labels for their images (not their annotations of course)?

mfojtik replied:
cool, so it is `io.k8s.` ?

bparees replied:
". Kube upstream uses docker suggested labels in the Kubelet (io.foo.bar)"

so yes, I guess so.

bparees commented on pkg/git/clone.go:
```diff
@@ -26,14 +27,13 @@ func (c *Clone) Download(config *api.Config) error {
-
 			if err := c.Checkout(targetSourceDir, config.Ref); err != nil {
-				return err
+				return c.GetInfo(targetSourceDir), err
```
errr...can't i have a context dir and a ref?

mfojtik replied:
yeah, seems like not... that is a bug in our code I guess.

mfojtik replied:
no, we return only if the checkout ref fails ;-) ufffff (and I just removed the GetInfo() if that happens)

bparees commented on pkg/git/git.go:
```diff
@@ -90,6 +93,28 @@ func (h *stiGit) Checkout(repo, ref string) error {
+		CommitID: git("rev-parse", "--short", "--verify", "HEAD"),
+		Author:   git("--no-pager", "show", "-s", "--format=%an <%ae>", "HEAD"),
+		Date:     git("--no-pager", "show", "-s", "--format=%ad", "HEAD"),
+		Message:  git("--no-pager", "show", "-s", "--format=%<(80,trunc)%s", "HEAD"),
```
no "..." if it exceeds 80 chars?  boo.

mfojtik replied:
the git truncate should add that suffix (at least I saw it on my machine ;-)

bparees replied:
ok

bparees commented on pkg/util/labels.go:
```diff
@@ -0,0 +1,40 @@
+	result[api.DefaultNamespace+"build.commit.id"] = info.CommitID
+	result[api.DefaultNamespace+"build.commit.ref"] = info.Ref
+	result[api.DefaultNamespace+"build.commit.message"] = info.Message
+	result[api.DefaultNamespace+"build.source-location"] = info.Location
```
might be better to check that you have a value for each field.  in the future we might support an SCM that has a subset of these fields.

bparees commented:
@mfojtik couple comments, one major concern which i realize wasn't introduced by you but would be good to fix here.

mfojtik commented:
@bparees updated, the current set of labels for the output image built by S2I now looks like this:

```json
"Labels": {
    "io.k8s.description": "Example description",
    "io.k8s.display-name": "My cool Ruby app",
    "io.openshift.s2i.build.commit.author": "Michal Fojtik \u003cmfojtik@redhat.com\u003e",
    "io.openshift.s2i.build.commit.date": "Wed Jun 24 13:36:57 2015 +0200",
    "io.openshift.s2i.build.commit.id": "bcbefa6",
    "io.openshift.s2i.build.commit.message": "initial",
    "io.openshift.s2i.build.commit.ref": "master-foo",
    "io.openshift.s2i.build.source-location": "file:///home/mfojtik/sandbox/ruby",
    "io.s2i.scripts-url": "image:///usr/local/sti",
    "k8s.io/description": "Platform for building and running Ruby 2.0 applications",
    "k8s.io/display-name": "Ruby 2.0",
    "openshift.io/expose-services": "8080:http",
    "openshift.io/tags": "builder,ruby,ruby20"
},
```

soltysh commented:
Still LGTM, nice work @mfojtik!

bparees commented:
@mfojtik I guess we then need to go back and  add "io.openshift.tags" and "io.openshift.expose-services" to our images so we can start to deprecate the "openshift.io" ones?

we can still convert those to "openshift.io/tags" when/if we make them annotations on the imagestreamtag.

bparees commented:
@mfojtik think you missed this comment?

    might be better to check that you have a value for each field. in the future we might support an SCM that has a subset of these fields.

bparees commented:
comment was not missed, I didn't read the addLabel function.
[merge]
